### PR TITLE
Homework

### PR DESCRIPTION
**Pull request by AusraSAFT**

### CODE REVIEW

File: homework/tests/calculatortest.spec.js [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+  '2',
+  '3',
+  '4',
+  '5',
+  '6',
+  '7',
+  '8',
+  '9'
+]
+
```
Comment (ricardassimcikasdb) [inline homework/tests/calculatortest.spec.js:16]: All the selectors in the code could described as constants if they do repeat. For example:

const firstNumberInputField = '#number1Field';

So then you can use it in code like if (await page.locator(firstNumberInputField ).isVisible())

Image situation if you have 2000 tests where selector is written in a way like you did, and it changes to something like '#numberXField' - now you would need to refactor code and change it in every single place instead of easily updating the selector constant.

File: homework/tests/calculatortest.spec.js [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+        await page.locator('#number1Field').type('6');
+        await page.locator('#number2Field').type('2');
+        await page.selectOption('#selectOperationDropdown', {label: 'Multiply'});
+        await page.locator('#calculateButton').click();
+    
+        await expect(page.locator('#numberAnswerField')).toHaveValue('12');
+      });
+    });
+  });
+
```
Comment (ricardassimcikasdb) [inline homework/tests/calculatortest.spec.js:47]: Let's take this last test "Dividing 6 and 2 results in 3" as an example for requirement: "**When test results in failure, actual result should be understandable from the test error message**.".

This timeout message "Test timeout of 10000ms exceeded." does not really provide a lot of information for us. When the test fail like this - we should check what is an issue, register it and improve the test so next time we will know that "Calculate" button is missing just looking into the report.

This could be achieved by using if() & else() blocks, for example:

```
if (await page.locator('#number1Field').isVisible()) { 
await page.locator('#number1Field').type('2'); } else{ 
throw "First number input field not working/missing"; }
```

Now if you would use that, message "First number input field not working/missing" would be displayed in the report, instead of timeout message. However, this is not a good practice to use those checks in every single test, so please update your code:

1. Try to exclude those checks into the separate function and call it in every test, if those checks fails - further test code will not be executed
2. If you will not be able to come up with solution described in 1st point, update only your test "Dividing 6 and 2 results in 3" with those checks.

image

File: homework/tests/calculatortest.spec.js [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    test.describe(version + ': First number input', () => {
+      test('Checking if first number input is visible', async ({ page }) => {
+        await page.goto('https://testsheepnz.github.io/BasicCalculator');
+        await page.selectOption('#selectBuild', { label: version});
+       
+        await expect(page.locator('#number1Field')).toBeVisible();
+      });
+    });
+  });
+
```
Comment (ricardassimcikasdb) [inline homework/tests/calculatortest.spec.js:230]: Those tests/checks where you check if dropdown/field/button is visible and displayed for the user could be removed as separate tests and other tests could be updated as in comment above:
image

File: homework/tests/calculatortest.spec.js [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+      await page.locator('#number1Field').type('2');
+      await page.locator('#number2Field').type('3');
+      await page.selectOption('#selectOperationDropdown', {label: 'Add'});
+      await page.locator('#calculateButton').click();
+  
+      await expect(page.locator('#numberAnswerField')).toHaveValue('5');
+    });
+  });
+});
+
```
Comment (ricardassimcikasdb) [inline homework/tests/calculatortest.spec.js:32]: It is good that you thought of adding integers and non-integers, but in the homework description it was described "**Tests should cover main functions, critical flows.**" , so that being said, only single test for adding numbers (integers or non-integers) would be enough without going into edge-cases. Please try to think about your automation strategy and remove tests which are repetitive (like adding integers and non-integers). 

However, let's leave 1 repetitive scenario. For this case, separate function should be created and called with different parameters. So instead of deleting all repetitive scenarios, please leave  2 tests regarding Add function and when you call those tests use function. I am attaching a not finished code snippet which I hope will help you to figure this out.

```
    async checkAddFunction(pageUrl, buildVersion, numberOne, numberTwo) {
        await page.goto(pageUrl);
        await page.selectOption('#selectBuild', {label: buildVersion});
        await page.locator('#number1Field').type(numberOne);
        ...
        ...
    }
```

```
 test('Adding -3 and 0 results in -3', async ({page}) => {
await checkAddFunction("https://testsheepnz.github.io/BasicCalculator", version, "3" ...);


            await expect(page.locator('#numberAnswerField')).toHaveValue('-3');
        });
    });
```

Disclaimer: I did not test this code and it might need some improvement, it is just for you to get an idea how it should be done.
